### PR TITLE
Adding frequentist numerical tests + Integrate StatsAPI

### PR DESCRIPTION
**Pull request by ShouvikGhosh2048**
This PR adds frequentist numerical tests, and also integrates StatsAPI into CRRao.

### CODE REVIEW

**Comment by ShouvikGhosh2048** [comment]
@sourish-cmi 
Currently this PR tests frequentist models on ordinary formulae and 0 intercept formulae. The tests just check whether the coeftables match.
- Should we test on other formulae/datasets?
- Should we test on more (or all) functions in src/frequentist/getter.jl?

**Comment by sourish-cmi** [comment]
1) We should test for  `predict`
2) Some possible formulas:
     i) `mpg~hp+hp^2 + wt + wt*hp`
     ii) `log(mpg)~log(hp)+log(wt)`
3) Later we can add few more `datasets`

**Comment by ShouvikGhosh2048** [comment]
> 1. We should test for  `predict`
> 
>     2. Some possible formulas:
>        i) `mpg~hp+hp^2 + wt + wt*hp`
>        ii) `log(mpg)~log(hp)+log(wt)`
> 
>     3. Later we can add few more `datasets`

I added predict and some formulae.

For Logistic, Negative binomial and Poisson should there be an example with log on the left?

**Comment by sourish-cmi** [comment]
> > 1. We should test for  `predict`
> >    
> >    2. Some possible formulas:
> >       i) `mpg~hp+hp^2 + wt + wt*hp`
> >       ii) `log(mpg)~log(hp)+log(wt)`
> >    3. Later we can add few more `datasets`
> 
> I added predict and some formulae.
> 
> For Logistic, Negative binomial and Poisson should there be an example with log on the left?

No @ShouvikGhosh2048  for Logistic, Negative binomial and Poisson, we don't need a log on the left.

File: src/frequentist/getter.jl [hunk 1]
```diff
@@ -221,7 +221,7 @@ function predict(container::FrequentistRegression{:LogisticRegression}, newdata:
         return Probit_Link.(z)
     elseif (container.link == GLM.CauchitLink)
         return Cauchit_Link.(z)
-    elseif (container.link == GLM.Cloglog)
+    elseif (container.link == GLM.CloglogLink)
```
[the remark this answers is not in the document]
Comment (sourish-cmi) [reply src/frequentist/getter.jl:224]: okay -- great. thanks.

**Comment by sourish-cmi** [comment]
Perhaps we should check for AIC and BIC

```R
> attach(mtcars)
> fit = lm(mpg ~ hp + wt + gear,mtcars)
> AIC(fit)
[1] 157.0528
> BIC(fit)
[1] 164.3815
```

Add it in the code and check if `CRRao` yields `AIC = 157.0528` and `BIC=164.3815`

**Comment by ShouvikGhosh2048** [comment]
@sourish-cmi 

```Julia
using RDatasets, GLM, StatsBase

mtcars = dataset("datasets", "mtcars")
glm_model = lm(@formula(MPG ~ HP + HP^2 + WT + WT * HP), mtcars)
print(aic(glm_model))
```

gives 145.92297178935536

and

```R
attach(mtcars)
fit = lm(mpg ~ hp + hp^2 + wt + wt * hp, mtcars)
AIC(fit)
```

gives 145.6109.

**Comment by sourish-cmi** [comment]
In your R-code, there is a syntax mistake. 
> ```r
> fit = lm(mpg ~ hp + hp^2 + wt + wt * hp, mtcars)
> AIC(fit)
> ```
The correct formula would be as follows:

```r
> fit = lm(mpg ~ hp + I(hp^2) + wt + wt * hp, mtcars)
> AIC(fit)
[1] 145.923
```

**Comment by ShouvikGhosh2048** [comment]
1)
```R
library("Zelig")
data(turnout)
fit = glm(vote ~ age + I(age^2) + race + income * educate, family = binomial(link = "cloglog"), data = turnout)
sprintf("%.15f", AIC(fit))
```
gives 2022.884413390652753,
while
```Julia
using GLM, RDatasets
turnout = dataset("Zelig", "turnout")
model = glm(@formula(Vote ~ Age + Age^2 + Race + Income * Educate), turnout, Binomial(), CloglogLink())
GLM.aic(model)
```
gives 2022.8844886368.

2)
```R
library("Zelig")
library(MASS)
data(sanction)
fit = glm.nb(num ~ target + coop + ncost, data = sanction)
sprintf("%.15f", AIC(fit))
```
gives 344.879549011806205,
while
```Julia
using GLM, RDatasets
sanction = dataset("Zelig", "sanction")
model = glm(@formula(Num ~ Target + Coop + NCost), sanction, NegativeBinomial(), LogLink())
GLM.aic(model)
```
gives 365.8580428654269.

**Comment by sourish-cmi** [comment]
> ```r
> library("Zelig")
> data(turnout)
> fit = glm(vote ~ age + I(age^2) + race + income * educate, family = binomial(link = "cloglog"), data = turnout)
> sprintf("%.15f", AIC(fit))
> ```
> 
> gives 2022.884413390652753, while
> 
> ```julia
> using GLM, RDatasets
> turnout = dataset("Zelig", "turnout")
> model = glm(@formula(Vote ~ Age + Age^2 + Race + Income * Educate), turnout, Binomial(), CloglogLink())
> GLM.aic(model)
> ```
> 
> gives 2022.8844886368.
> 
> 
> ```r
> library("Zelig")
> library(MASS)
> data(sanction)
> fit = glm.nb(num ~ target + coop + ncost, data = sanction)
> sprintf("%.15f", AIC(fit))
> ```
> 
> gives 344.879549011806205, while
> 
> ```julia
> using GLM, RDatasets
> sanction = dataset("Zelig", "sanction")
> model = glm(@formula(Num ~ Target + Coop + NCost), sanction, NegativeBinomial(), LogLink())
> GLM.aic(model)
> ```
> 
> gives 365.8580428654269.

Okay, it is an issue with `GLM.jl`. Can you please raise an issue in `GLM.jl`?

**Comment by ayushpatnaikgit** [comment]
> 

@mousum-github can you please check this?

**Comment by mousum-github** [comment]
image

**Comment by mousum-github** [comment]
`glm.nb` estimates the `\theta`  but glm with a Negative Binomial, you have to give theta and the default value is `0.5`.
`negbin` is the corresponding function where `\theta` is estimated from the data.

**Comment by ShouvikGhosh2048** [comment]
@mousum-github 
1) For the Julia program in (2), what will the corresponding R program be?
2) Do the programs in (1) match in what they should compute? If so, is the difference significant?

**Comment by mousum-github** [comment]
> @mousum-github
> 
> 1. For the Julia program in (2), what will the corresponding R program be?
    `Not sure, you may try with negative binomial distribution as a family parameter in R glm function`
> 2. Do the programs in (1) match in what they should compute? If so, is the AIC difference significant?
   `model = glm(@formula(Vote ~ Age + Age^2 + Race + Income * Educate), turnout, Binomial(), CloglogLink(); rtol=1.0E-16)`
   `GLM.aic(model)`
   `2022.8844131094252`

**Comment by sourish-cmi** [comment]
@ShouvikGhosh2048 for now, we can remove this example with `cloglog`.  We can say on the front page that the test for `cloglog` link function is not completed yet. We have to move ahead with other tests.  Particularly with the Bayesian test.

In this [issue](https://github.com/xKDR/CRRao.jl/issues/52) can you mention the completeness of the tests? or the cases that I have not considered yet !!

**Comment by ShouvikGhosh2048** [comment]
> @ShouvikGhosh2048 for now, we can remove this example with `cloglog`. We can say on the front page that the test for `cloglog` link function is not completed yet. We have to move ahead with other tests. Particularly with the Bayesian test.
> 
> In this [issue](https://github.com/xKDR/CRRao.jl/issues/52) can you mention the completeness of the tests? or the cases that I have not considered yet !!

1) I couldn't construct a R program for NegativeBinomial, so those tests are still incomplete.
2) I can start working on Bayesian tests parallelly.

**Comment by sourish-cmi** [comment]
@ShouvikGhosh2048 @mousum-github 

As discussed in yesterday's meeting we will focus on functional testing. 

1) @ShouvikGhosh2048  will develop the functional testing for the Bayesian module. For a given seed, the Bayesian module should reproduce the result.

2) @mousum-github will correct the negative binomial model with `negbin` instead of `glm` and provide the test with `R` output

**Comment by sourish-cmi** [comment]
@ShouvikGhosh2048 Mousum da added the negative binomial test, and I approved and merged the test. Perhaps you `pull` the latest from the main branch. Please check.

**Comment by sourish-cmi** [comment]
@ShouvikGhosh2048 since you created separate [PR](https://github.com/xKDR/CRRao.jl/pull/69) for Bayesian tests. I believe now we can merge this PR.  Please confirm.

**Comment by ShouvikGhosh2048** [comment]
I've updated the numerical tests for negative binomial. I haven't matched the AIC/BIC values with R - I'm not sure which values to pass to R's negbin to match Julia's negbin. 

I tried:
```R
library(aod)
library(Zelig)
data(sanction)
fit = negbin(num ~ 0 + target + coop + ncost, ~1, sanction)
AIC(fit)
```
This gives 344.896 compared to CRRao's 344.87955007518536.
Otherwise the PR is ready to be merged.

**Comment by ShouvikGhosh2048** [comment]
We can merge it for now (to make sure our frequentist models are consistent), but we should keep track of the difference in Negative Binomial models.

**Comment by sourish-cmi** [comment]
I have filed the [issue](https://github.com/xKDR/CRRao.jl/issues/70)  with the title **AIC value for negbin regression does not matches with `R`**. Otherwise, this PR is fine - so I am merging the PR with the main.

**Review by sourish-cmi — APPROVED** [review]
All codes are checked and passed. Closing the PR - while opening  a new [issue](https://github.com/xKDR/CRRao.jl/issues/70)

**Pull request merged**